src/app/reporting/Engine.cpp

```diff
@@ -47,6 +47,18 @@ void Engine::Shutdown()
     mGlobalDirtySet.ReleaseAll();
 }
 
+bool Engine::IsClusterVersionAllowed(ClusterInfo * aDataVersionFilterList, const ConcreteReadAttributePath & aPath)
+{
+    for (auto filter = aDataVersionFilterList; filter != nullptr; filter = filter->mpNext)
+    {
+        if (aPath.mEndpointId == filter->mEndpointId && aPath.mClusterId == filter->mClusterId)
+        {
+            return IsClusterDataVersionAllowed(filter->mEndpointId, filter->mClusterId, filter->mDataVersion);
```

Comment:
What if there were multiple data versions for the same endpoint+cluster in the list?  What does the spec say?

Reply:
spec does not mention multiple data versions for the same endpoint+cluster.  I expect controller would not send data versions for the same endpoint+cluster, maybe I should do dup check here?

Reply:
Spec does not disallow sending the same endpoint+cluster multiple times with different data versions.... What the spec says is:

> If the DataVersionFilters field is present and does not indicate a DataVersionFilterIB entry with a Path field that matches the cluster path, or the DataVersion field of a matching DataVersionFilterIB does not match the data version of the cluster instance in the path, then an AttributeDataIB SHALL be generated with the attribute data for the path and cluster data version

In particular, if you send the same path with multiple data versions you will get the data back per the spec (because at least one of those will fail to match).

This is actually quite critical for something that is multiplexing multiple separate subscriptions in some way.

That does not seem to be the behavior implemented here, right?  In terms of `IsClusterDataVersionMatch`, this seems like it should return false if either nothing in the list matches the given endpoint+cluster in the path or there is an entry in the list that matches the endpoint+cluster in the path but does not match the current data version of that cluster.  Which in particular means we always have to walk the full list, without taking any early returns.

Reply:
understood, agree, thanks

Reply:
understood, thanks for spec clarification on this.